File: mobile/android/app/src/androidTest/java/io/meshtech/koasta/extras/StubApi.kt
package io.meshtech.koasta.extras

import io.meshtech.koasta.net.ApiLegalDocument
import io.meshtech.koasta.net.IApi
import io.meshtech.koasta.net.model.*

class StubApi: IApi {
  var loginResult = ApiPlainResult(null)
  var getVenuesResult = ApiResult<List<Venue>>(null, null)
  var queryVenuesResult = ApiResult<List<Venue>>(null, null)
  var getMenusResult = ApiResult<List<Menu>>(null, null)
  var getVenueResult = ApiResult<Venue>(null, null)
  var sendOrderResult = ApiResult<SendOrderResult>(null, null)
  var requestOrderEstimateResult = ApiResult<EstimateOrderResult>(null, null)
  var getOrdersResult = ApiResult<List<HistoricalOrder>>(null, null)
  var createDeviceResult = ApiResult<Void>(null, null)
  var getOrderResult = ApiResult<HistoricalOrder>(null, null)
  var fetchLegalDocumentResult = ApiResult<String>(null, null)
  var submitReviewResult = ApiResult<Void>(null, null)

  override suspend fun login(): ApiPlainResult = loginResult
  override suspend fun getVenues(lat: Double?, lon: Double?): ApiResult<List<Venue>> = getVenuesResult
  override suspend fun queryVenues(str: String): ApiResult<List<Venue>> = queryVenuesResult
  override suspend fun getMenus(venueId: Int): ApiResult<List<Menu>> = getMenusResult
  override suspend fun getVenue(venueId: Int): ApiResult<Venue> = getVenueResult
  override suspend fun sendOrder(order: Order): ApiResult<SendOrderResult> = sendOrderResult
  override suspend fun requestOrderEstimate(order: DraftOrder): ApiResult<EstimateOrderResult> = requestOrderEstimateResult
  override suspend fun getOrders(): ApiResult<List<HistoricalOrder>> = getOrdersResult
  override suspend fun createDevice(token: String): ApiResult<Void> = createDeviceResult
  override suspend fun getOrder(id: Int): ApiResult<HistoricalOrder> = getOrderResult
  override suspend fun fetchLegalDocument(type: ApiLegalDocument): ApiResult<String> = fetchLegalDocumentResult
  override suspend fun submitReview(id: Int, review: VenueReview): ApiResult<Void> = submitReviewResult
}

File: mobile/ios/test/ui/stubs/stub-api.swift
import Foundation
import Hydra
@testable import pubcrawl


class StubApi: Api {
  var getVenueResult: Venue? = nil
  var getMenuListResult: [Menu]? = nil
  var getVenuesResult: [Venue]? = nil
  var getVenueShouldThrow = false
  var getEphemeralKeyResult: AuthEphemeralKeyResult? = nil
  var getNearbyVenuesResult: [Venue]? = nil
  var sendOrderResult: SendOrderResult? = nil
  var getOrdersResult: [HistoricalOrder]? = nil
  var getOrderResult: HistoricalOrder? = nil
  var getEstimateResult: EstimateOrderResult? = nil
  var getProfileResult: UserProfile = UserProfile(registrationId: "a", wantAdvertising: true, votedVenueIds: [123], firstName: nil, lastName: nil)

  var getMenuListShouldThrow = false
  var loginResultShouldThrow = false
  var getVenuesShouldThrow = false
  var getEphemeralKeyShouldThrow = false
  var getNearbyVenuesShouldThrow = false
  var sendOrderShouldThrow = false
  var getOrdersShouldThrow = false
  var getOrderShouldThrow = false
  var getEstimateShouldThrow = false
  var submitReviewShouldThrow = false
  var getProfileShouldThrow = false

  func login(firstName: String?, lastName: String?) -> Promise<Void> {
    return async({ _ in
      if self.loginResultShouldThrow {
        throw ApiError(statusCode: 500, body: nil, bodyData: nil)
      }
    })
  }

  func getNearbyVenues(lat: Double, lon: Double, limit: Int) -> Promise<[Venue]?> {
    return async({ _ in
      if self.getNearbyVenuesShouldThrow {
        throw ApiError(statusCode: 500, body: nil, bodyData: nil)
      }

      return self.getNearbyVenuesResult
    })
  }

  func getVenues(query: String) -> Promise<[Venue]?> {
    return async({ _ in
      if self.getVenuesShouldThrow {
        throw ApiError(statusCode: 500, body: nil, bodyData: nil)
      }

      return self.getVenuesResult
    })
  }

  func getPaymentProviderEphemeralKey() -> Promise<AuthEphemeralKeyResult> {
    return async({ _ in
      if self.getEphemeralKeyShouldThrow {
        throw ApiError(statusCode: 500, body: nil, bodyData: nil)
      }

      return self.getEphemeralKeyResult!
    })
  }

  func sendOrder(_ order: Order) -> Promise<SendOrderResult> {
    return async({ _ in
      if self.sendOrderShouldThrow {
        throw ApiError(statusCode: 500, body: nil, bodyData: nil)
      }

      return self.sendOrderResult!
    })
  }

  func getOrders() -> Promise<[HistoricalOrder]> {
    return async({ _ in
      if self.getOrdersShouldThrow {
        throw ApiError(statusCode: 500, body: nil, bodyData: nil)
      }

      return self.getOrdersResult!
    })
  }

  func getOrder(orderId: Int) -> Promise<HistoricalOrder> {
    return async({ _ in
      if self.getOrderShouldThrow {
        throw ApiError(statusCode: 500, body: nil, bodyData: nil)
      }

      return self.getOrderResult!
    })
  }

  func fetchPrivacyPolicy() -> Promise<String> {
    return Promise(resolved: "abc")
  }

  func fetchTermsAndConditions() -> Promise<String> {
    return Promise(resolved: "def")
  }

  func fetchCurrentFlags() -> Promise<FeatureFlags> {
    return Promise(resolved: FeatureFlags(flags: FeatureFlagListing(facebookAuth: true)))
  }


  func registerPushToken(_ registration: PushRegistration) -> Promise<Void> {
    return Promise(resolved: ())
  }

  func getVenue(forId id: String) -> Promise<Venue?> {
    return async({ _ in
      if self.getVenueShouldThrow {
        throw ApiError(statusCode: 500, body: nil, bodyData: nil)
      }

      return self.getVenueResult
    })
  }

  func getMenuList(forId id: String) -> Promise<[Menu]?> {
    return async({ _ in
      if self.getMenuListShouldThrow {
        throw ApiError(statusCode: 500, body: nil, bodyData: nil)
      }

      return self.getMenuListResult
    })
  }

  func submitReview(forId id: String, review: VenueReview) -> Promise<Void> {
    return async({ _ in
      if self.submitReviewShouldThrow {
        throw ApiError(statusCode: 500, body: nil, bodyData: nil)
      }
    })
  }

  func requestOrderEstimate(_ order: DraftOrder) -> Promise<EstimateOrderResult> {
    return async({ _ in
      if self.getEstimateShouldThrow {
        throw ApiError(statusCode: 500, body: nil, bodyData: nil)
      }

      return self.getEstimateResult!
    })
  }

  func fetchCurrentProfile() -> Promise<UserProfile> {
    return async({ _ in
      if self.getProfileShouldThrow {
        throw ApiError(statusCode: 401, body: nil, bodyData: nil)
      }

      return self.getProfileResult
    })
  }
}
